app/src/main/java/com/garethevans/church/opensongtablet/StageMode.java

```diff
@@ -7146,24 +7233,42 @@ protected void onPostExecute(String s) {
                         prepareView();
                     }
 
-                    // Automatically stop the metronome if it is still on
-                    if (StaticVariables.metronomeonoff.equals("on")) {
-                        // Stop it
-                        Metronome.startstopMetronome(StageMode.this, StageMode.this,
-                                preferences.getMyPreferenceBoolean(StageMode.this,"metronomeShowVisual",false),
-                                defmetronomecolor, preferences.getMyPreferenceString(StageMode.this,"metronomePan","C"),
-                                preferences.getMyPreferenceFloat(StageMode.this,"metronomeVol",0.5f),
-                                preferences.getMyPreferenceInt(StageMode.this,"metronomeLength",0));
```

Comment:
I wonder if this is the cause of the metronome issue on moving between songs?

Reply:
I had though I had change to gesture for all metronome calls - have I missed off a change!  Let me look.

Reply:
I have changed over to gesture7 use...  I  do not see this stop.  So would need to understand the metronome settings you are using for the original and new song to try to replicate.  What are the settings?
Are you testing against a full merge?

Reply:
Hi Ian,
Yes, I created a new pull request from your fort to a branch named IanTest.  I accepted this extra pull request so I could test without affecting the master branch just now.
Have a song with metronome values set.  Start the metronome (either using the popup start button or long pressing).
Then try to swipe to the next song.  Two issues:
1.  If you are in a set, nothing appears to happen when you swipe.  As soon as you stop the metronome, the swipe is called.
2.  If you are not in a set, the song is swiped away, but the new one isn't loaded (black screen) until you stop the metronome.

Additionally, the metronome should be stopped if it hasn't been set for the newly loaded song after swiping (although this bit was also an issue in my code).

I think this is the only issue to resolve - really well done on figuring out my ugly, messy code and making it better!

Reply:
Mess is much better than nothing!  I can work with moderate mess.  Just a few tweaks to assist.

The code stops a playing old song metronome using gesture7 and then, if needed, gesture7 to start the new songs metronome.  So problem is in gesture7 and supporting code.

                    // Do not touch on a reload
                    if (!StaticVariables.reloadOfSong) {
                        // Stop it
                        // IV - gesture used for stop and start
                        if (StaticVariables.clickedOnMetronomeStart) {
                            gesture7();
                            // Metronome Was playing - if requested autostart start the metronome for a new song
                            if (preferences.getMyPreferenceBoolean(StageMode.this, "metronomeAutoStart", false) &&
                                    FullscreenActivity.isSong) {
                                // Start it
                                gesture7();
                            }
                        }
                    }

Perhaps you can debug to see what is going on?  Perhaps StaticVariables.clickedOnMetronomeStart is not being set correctly elsewhere.

I do the test.  In a set and move to a song with metronome OK.  Move to song with not set metronome gives toast.   What are your detailed metronome settings please.

Reply:
That's weird that it is working in yours.  I'm going to try a reset/invalidate of the caches in Android Studio as the code is showing me some errors, yet still running - something isn't right at my end!

Reply:
OK

Reply:
Hang on... I seem to remember that during testing the emulator struggles with the metronome code.  Try a build and run on a device.  Also try your original code on the emulator to see (hopefully!) if it also mis-behaves.

Reply:
Just tried here.  Yes, the emulator does not do well!  Slow response and the occasional crash and only moves song when metronome stopped.  No problems on Fire HD 10 or Moto G6 Play.   Did not investigate the cause.

Reply:
I was reporting it working for my Fire HD 10.

Reply:
I program on Linux and can't use the emulator!  My test device is a Samsung Galaxy Nexus 10, which despite being old (and updated to Marshmallow using custom ROM), works really well.  Better performance than my newer Samsung Galaxy Tab 10.  I'll continue to debug this jut now.

Reply:
OK.  There has been a slight reorg of code so the stop code is called by gesture 7 - but the call appears to be no different.  Strange.

Reply:
Thanks for letting me know your dev env.  FYI.  I am Windows 10 / AS 4.0.1 with devices emulator, Fire 10 HD, Fire 8  and Moto 6 Play.  This is my first coding using AS and Android/Java.  So thanks for giving me a good reason to get into a new language.

Reply:
Done some testing.  If I stop the metronome just before starting the loadsong_async task the song loads fine, but the metronome isn't resuming on the next song.  If I leave the start stop logic to be run as part of the async task itself, the async task never starts until I stop the metronome.  Very strange - I wonder if there is a limit to the number of tasks that can run simultaneously on some devices (since you don't see this issue).
I'm heading out just now, but at least I've figured out a potential fix - split the following bit up into a stop (before load) and  restart if required after load:

if (!StaticVariables.reloadOfSong) {
                            // Stop it
                            // IV - gesture used for stop and start
                            if (StaticVariables.clickedOnMetronomeStart) {
                                gesture7();
                                // Metronome Was playing - if requested autostart start the metronome for a new song
                                if (preferences.getMyPreferenceBoolean(StageMode.this, "metronomeAutoStart", false) &&
                                        FullscreenActivity.isSong) {
                                    // Start it
                                    gesture7();
                                }
                            }
                        }

Reply:
Sounds like a way forward.  Great.

Reply:
Am I right in thinking metronome is not called async like other actions?  Perhaps it should be to stop any interference / delay to song display?

Reply:
My feeling is to async metronome and leave the call as-is. So do what you have done with the other actions.  This will keep the logic very clear (no splitting of activity) and get out of the way of song load (which is good).  This has the advantage of keeping the approach consistent across actions. Kind regards. Ian.

Reply:
Yeah, the metronome is already Async - initiated and run from the Metronome.java class.  It is actually split into two async tasks - one for the metronome, which doesn't need a UI update, so all done in doInBackground task and one that handles the visual metronome, which flashes the actionbar, so does the logic in a doInBackground task and the UI update in OnProgressUpdate,.  They used to be merged (using if... statements to decide which bits to fire), but I definitely felt performance was more solid (less timing issues) when split up and called individually.

Reply:
Ok, I've tracked my way through the app using logcat statements and found that the app wouldn't get past the animateInSong without the metronome being stopped first.  Not quite sure why!  So, I've moved the restart gesture7 logic to after the animateIn (which makes sense as displaying the song should be the highest priority in the chain).  It all works now!!!

I'm just going to make a few other changes (to address Android 8+ devices showing blank storage locations when using the file picker library), then I'll merge the pull requests to the master and get you to update for testing at your end.

Reply:
Ahh.  Should have looked harder!  Leave it with you.  Have fun!

Reply:
Great stuff. 

Just a note for testing... i have focused on songs so not PDF or image which I know are supported.  For me they are not good song types so I do not use but please do make sure to regression test (there is lots of things I don't do).

Thanks for being so receptive to the new features of the commit.